pkg/agent/controller/traceflow/traceflow_controller_test.go

```diff
@@ -0,0 +1,577 @@
+		ofClient:              mockOFClient,
+		networkPolicyQuerier:  npQuerier,
+		ovsBridgeClient:       ovsClient,
+		interfaceStore:        ifaceStore,
+		networkConfig:         networkConfig,
+		nodeConfig:            nodeConfig,
+		serviceCIDR:           serviceCIDRNet,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "traceflow"),
+		runningTraceflows:     make(map[uint8]*traceflowState),
+	}
+	return &fakeTraceflowController{
+		Controller:         tfController,
+		kubeClient:         kubeClient,
+		mockController:     controller,
+		mockOFClient:       mockOFClient,
+		crdClient:          crdClient,
+		crdInformerFactory: crdInformerFactory,
+		ovsClient:          ovsClient,
+	}
+}
+
+func addPodInterface(ifaceStore interfacestore.InterfaceStore, podNamespace, podName, podIP, podMac string, ofPort int32) {
+	containerName := k8s.NamespacedName(podNamespace, podName)
+	ifIPs := make([]net.IP, 0)
+	ifIPs = append(ifIPs, net.ParseIP(podIP))
+	mac, _ := net.ParseMAC(podMac)
+	ifaceStore.AddInterface(&interfacestore.InterfaceConfig{
+		IPs:                      ifIPs,
+		MAC:                      mac,
+		InterfaceName:            util.GenerateContainerInterfaceName(podName, podNamespace, containerName),
+		ContainerInterfaceConfig: &interfacestore.ContainerInterfaceConfig{PodName: podName, PodNamespace: podNamespace, ContainerID: containerName},
+		OVSPortConfig:            &interfacestore.OVSPortConfig{OFPort: ofPort},
+	})
+}
+
+func TestPreparePacket(t *testing.T) {
+	tcs := []struct {
+		name           string
+		tf             *crdv1alpha1.Traceflow
+		receiverOnly   bool
+		expectedPacket *binding.Packet
+		expectedErr    string
+	}{
+		{
+			name: "invalid destination ipv4",
+			tf: &crdv1alpha1.Traceflow{
+				ObjectMeta: metav1.ObjectMeta{Name: "tf1", UID: "uid1"},
+				Spec: crdv1alpha1.TraceflowSpec{
+					Source: crdv1alpha1.Source{
+						Namespace: "default",
+						Pod:       "pod-1",
+					},
+					Destination: crdv1alpha1.Destination{
+						IP: "1.1.1.300",
+					},
+				},
+			},
+			expectedErr: "invalid destination IP address",
+		},
+		{
+			name: "empty destination with no live traffic",
+			tf: &crdv1alpha1.Traceflow{
+				ObjectMeta: metav1.ObjectMeta{Name: "tf2", UID: "uid2"},
+				Spec: crdv1alpha1.TraceflowSpec{
+					Source: crdv1alpha1.Source{
+						Namespace: "default",
+						Pod:       "pod-1",
+					},
+				},
+			},
+			expectedErr: "destination is not specified",
+		},
+		{
+			name: "receive only from a source ipv4 in live traffic traceflow",
+			tf: &crdv1alpha1.Traceflow{
+				ObjectMeta: metav1.ObjectMeta{Name: "tf3", UID: "uid3"},
+				Spec: crdv1alpha1.TraceflowSpec{
+					Source: crdv1alpha1.Source{
+						IP: "192.168.12.4",
+					},
+					LiveTraffic: true,
+				},
+			},
+			receiverOnly: true,
+			expectedPacket: &binding.Packet{
+				SourceIP:       net.ParseIP("192.168.12.4"),
+				DestinationMAC: pod1MAC,
```

Comment:
It's not clear why the above traceflow could get this packet, I think it needs to specify destination Pod.